src/main/scala/org/apache/spark/ml/linalg/distributed/RapidsRowMatrix.scala

```diff
@@ -88,17 +88,29 @@ class RapidsRowMatrix(
     if (useCuSolverSVD) {
       val nvtxRangeSVD = new NvtxRange("cuSolver SVD", NvtxColor.BLUE)
 
-      val dense_U = DenseMatrix.zeros(n, n)
 
-      val dense_S = DenseMatrix.zeros(1, n)
+      var svdOutput: Array[(DenseMatrix, DenseMatrix)] = null
+      val gpuIdBC = rows.context.broadcast(gpuId)
       try {
-        // this is done on driver, so no task resources here, assign 0 manually.
-        RAPIDSML.calSVD(n, Cov, dense_U, dense_S, 0)
+        val rddTmp = rows.context.parallelize(Seq(0), 1)
+
+        val svdRdd = rddTmp.mapPartitions( _ => {
+          val gpu = if (gpuIdBC.value == -1) {
+            TaskContext.get().resources()("gpu").addresses(0).toInt
```

Comment:
This will throw exceptions in local mode.

Try

  def getGPUAddrFromResources: Int = {
    val tc = TaskContext.get()
    if (tc == null) {
      throw new RuntimeException("Something wrong for task context")
    }
    val resources = tc.resources()
    if (resources.contains("gpu")) {
      val addrs = resources("gpu").addresses
      if (addrs.size > 1) {
        // TODO should we throw exception ?
        logger.warn("XGBoost only supports 1 gpu per worker")
      }
      // take the first one
      addrs.head.toInt
    } else {
      throw new RuntimeException("gpu is not allocated by spark, " +
        "please check if gpu scheduling is enabled")
    }
  }

Reply:
We have a dedicated config for this case `PCA().setGpuId()` when we see the exception like "resource no key for gpu".
It exists on both training and transform. I think I can apply this fix for it in another follow-up PR for better UX.